### PR TITLE
virts-487b

### PR DESCRIPTION
**Pull request by ArtificialErmine**
Adds the ability to specify specific, as well as general (default) responses for abilities, based on the variables used in the commands issued.

### CODE REVIEW

File: app/c_simulation.py [hunk 1]
```diff
@@ -12,11 +12,13 @@ def display(self):
         return dict(name=self.name, ability_id=self.ability_id, paw=self.paw, status=self.status,
                     response=self.response)
```
Comment (khyberspache) [inline app/c_simulation.py]: Need to add variable name and value to the display property?

File: app/c_simulation.py [hunk 2]
```diff
@@ -12,11 +12,13 @@ def display(self):
...
 
-    def __init__(self, name, ability_id, paw, status, response):
+    def __init__(self, name, ability_id, paw, status, response, variable_name=None, variable_value=None):
         self.name = name
         self.ability_id = ability_id
         self.paw = paw
         self.status = status
+        self.v_name = variable_name
+        self.v_value = variable_value
```
Comment (khyberspache) [inline app/c_simulation.py]: For consistency should we just keep the internal variables `variable_name` `variable_value`?

File: app/simulation_svc.py [hunk 3]
```diff
@@ -98,3 +101,32 @@ def __init__(self, services, agents, loaded_scenario):
...
                 return True
         return False
+
+    """ PRIVATE"""
+
+    async def _extract_vars(self, ability, link):
+        decoded_test = self.decode_bytes(ability.test)
+        decoded_command = self.decode_bytes(link.command)
+        variables = re.findall(r'#{(.*?)}', decoded_test, flags=re.DOTALL)
+        vars = [x for x in variables if x not in ['server', 'paw', 'location', 'group']]
```
Comment (khyberspache) [inline app/simulation_svc.py]: vars() is a reserved function name in python... while this isn't a function call obviously, we should avoid using close keywords/key functions (same as what we did with prop/property)

File: app/simulation_svc.py [hunk 4]
```diff
@@ -98,3 +101,32 @@ def __init__(self, services, agents, loaded_scenario):
...
+        search_set = []
+        filtering = decoded_test
+        for v in vars:
+            search_key, filtering = filtering.split("#{" + v + "}", 1)
```
Comment (khyberspache) [inline app/simulation_svc.py]: Single quotes throughout

File: app/simulation_svc.py (hunk 3 above)
Comment (khyberspache) [reply app/simulation_svc.py]: You're only using vars in one place in the function - you can optimize this by just eliminating the variable:

`for v in [x for x in variables if x not in ['server', 'paw', 'location', 'group']]`

File: app/simulation_svc.py [hunk 5]
```diff
@@ -98,3 +101,32 @@ def __init__(self, services, agents, loaded_scenario):
...
+            extract = re.findall(r'' + re.escape(search_key) + '(.*)', decoded_command, flags=re.DOTALL)
+            extract = extract[0]
```
Comment (khyberspache) [inline app/simulation_svc.py]: should be able to combine with the above:
`extract = (re.findall(r'' + re.escape(search_key) + '(.*)', decoded_command, flags=re.DOTALL))[0]`

File: app/simulation_svc.py [hunk 6]
```diff
@@ -98,3 +101,32 @@ def __init__(self, services, agents, loaded_scenario):
...
+                bound = 3 if len(filtering) > 2 else -1
+                extract = extract.split(filtering[0:bound])[0]
+            search_set.append([v, extract])
+        return search_set
+
+    async def _target_response(self, vars, response_object):
+        computable = dict()
+        for entry in response_object:
+            computable[entry.v_name] = {entry.v_value: [entry.response, entry.status]}
+        for v in vars:
```
Comment (khyberspache) [inline app/simulation_svc.py]: replace vars throughout per the above comment

File: app/simulation_svc.py [hunk 7]
```diff
@@ -98,3 +101,32 @@ def __init__(self, services, agents, loaded_scenario):
...
+            if v[0] in computable:
+                if v[1] in computable[v[0]]:
+                    return computable[v[0]][v[1]]
+        return computable['default']['default']
```
Comment (khyberspache) [inline app/simulation_svc.py:127]: Can you explain this to me a bit more?  Is the v_name and v_value always going to be Default?

**Review by khyberspache — CHANGES_REQUESTED** [review]
Overall good code.  Had a first past on the easy items with some questions.  Once revised I will test.

File: app/c_simulation.py (hunk 1 above)
Comment (ArtificialErmine) [reply app/c_simulation.py]: Easily done.

File: app/simulation_svc.py (hunk 3 above)
Comment (ArtificialErmine) [reply app/simulation_svc.py]: Easily fixed.

File: app/simulation_svc.py (hunk 5 above)
Comment (ArtificialErmine) [reply app/simulation_svc.py]: I could have sworn I put that back as one line (I broke it up for debugging at some point).

File: app/c_simulation.py (hunk 2 above)
Comment (ArtificialErmine) [reply app/c_simulation.py]: Yeah, that makes sense.

File: app/simulation_svc.py (hunk 4 above)
Comment (ArtificialErmine) [reply app/simulation_svc.py]: Woops.

File: app/simulation_svc.py (hunk 6 above)
Comment (ArtificialErmine) [reply app/simulation_svc.py]: Taken care of.

**Review by khyberspache — CHANGES_REQUESTED** [review]
I was able to get the first simulation (hunter & simulation group) to run, but I wasn’t able to run the Hypervisor attack with the hypervisor group.  Looks like there might be UI related issue with actually applying the active simulation?

File: app/simulation_svc.py [hunk 8]
```diff
@@ -98,3 +101,28 @@ def __init__(self, services, agents, loaded_scenario):
                 await self.start_agent(agent)
                 return True
         return False
+
+    async def _extract_used(self, ability, link):
+        decoded_test = self.decode_bytes(ability.test)
+        decoded_command = self.decode_bytes(link.command)
+        variables = re.findall(r'#{(.*?)}', decoded_test, flags=re.DOTALL)
+        search_set = []
+        filtering = decoded_test
```
Comment (khyberspache) [inline app/simulation_svc.py]: Why are we redeclaring this?

File: app/c_simulation.py [hunk 9]
```diff
@@ -10,13 +10,15 @@ def unique(self):
...
-                    response=self.response)
+                    response=self.response, variable_name=self.variable_name, variable_value=self.variable_value)
 
-    def __init__(self, name, ability_id, paw, status, response):
+    def __init__(self, name, ability_id, paw, status, response, variable_name=None, variable_value=None):
         self.name = name
         self.ability_id = ability_id
         self.paw = paw
         self.status = status
+        self.variable_name = variable_name
```
Comment (khyberspache) [inline app/c_simulation.py]: I couldn't figure out why this was sticking out to me and just did... we should be consistent with the Fact() object and other references throughout the application with regards to facts:
`self.variable_trait = variable_trait`

File: app/simulation_svc.py (hunk 8 above)
Comment (ArtificialErmine) [reply app/simulation_svc.py]: Fixed.

File: app/c_simulation.py (hunk 9 above)
Comment (ArtificialErmine) [reply app/c_simulation.py]: Updated it here and where applicable elsewhere.

**Pull request merged**